pkg/build/api/v1/types.go

```diff
@@ -41,19 +52,72 @@ type BuildSpec struct {
 	// output describes the Docker image the Strategy should produce.
 	Output BuildOutput `json:"output,omitempty"`
 
-	// resource requirements to execute the build
+	// resources computes resource requirements to execute the build.
 	Resources kapi.ResourceRequirements `json:"resources,omitempty"`
 
 	// postCommit is a build hook executed after the build output image is
 	// committed, before it is pushed to a registry.
 	PostCommit BuildPostCommitSpec `json:"postCommit,omitempty"`
 
-	// completionDeadlineSeconds is an optional duration in seconds, counted from the time when a build pod gets
-	// scheduled in the system, that the build may be active on a node before the
-	// system actively tries to terminate the build; value must be positive integer
+	// completionDeadlineSeconds is an optional duration in seconds, counted from
+	// the time when a build pod gets scheduled in the system, that the build may
+	// be active on a node before the system actively tries to terminate the
+	// build; value must be positive integer
 	CompletionDeadlineSeconds *int64 `json:"completionDeadlineSeconds,omitempty"`
 }
 
+// BuildTriggerCause holds information about a triggered build. It is used for
+// displaying build trigger data for each build and build configuration in oc
+// describe. It is also used to describe which triggers led to the most recent
+// update in the build configuration.
+type BuildTriggerCause struct {
+	// message is used to store a human readable message for why the build was
+	// triggered. E.g.: "Manually triggered by user", "Configuration change",etc.
+	Message string `json:"message,omitempty"`
+
+	// genericWebHook holds data about a builds generic webhook trigger.
+	GenericWebHook *GenericWebHookCause `json:"genericWebHook,omitempty"`
+
+	// gitHubWebHook represents data for a GitHub webhook that fired a
+	//specific build.
+	GitHubWebHook *GitHubWebHookCause `json:"githubWebHook,omitempty"`
+
+	// imageChangeBuild stores information about an imagechange event
+	// that triggered a new build.
+	ImageChangeBuild *ImageChangeCause `json:"imageChangeBuild,omitempty"`
+}
+
+// GenericWebHookCause holds information about a generic WebHook that
+// triggered a build.
+type GenericWebHookCause struct {
+	// revision is the git source revision information of the trigger.
+	Revision *SourceRevision `json:"revision,omitempty"`
```

Comment:
Please indicate in the comments which of these fields are optional (for all four of these structs)